### ISSUE TITLE
Wrong value stored when itemset changes during Ajax request

### ISSUE THREAD

**Issue by avernet** [orbeon/orbeon-forms#3038]
To reproduce, run 
[this form](https://github.com/orbeon/orbeon-forms/files/647400/form.txt) as stand-alone XForms, use an app like Charles to throttle the connection, with Firefox load the form, click on the dropdrown to open it, hit the down key, wait for the Ajax query to be sent and before the response arrives hit tab. Notice how *Orange* shows in dropdown (correctly), while the `xf:output` shows *Apple* (incorrectly).

This happens as the itemset changes as a consequence of the first Ajax request, and while the Ajax request is processed by the server, the client queues an event with a position in the "old itemset", and this event is sent as a second Ajax request, after the itemset change, so the server takes it to point to a different item than the one intended by the client when it queued the request.

A solution to this problem would be to use, instead of a position in the itemset, a hash of the item value. If we were to use this technique, we would have to take in consideration the possible impact on performance hashing would have.

[+1 from customer](https://3.basecamp.com/3600924/buckets/2021675/messages/317021140)

**Comment by ebruchez** [comment]
Ideally, hashes would be as opaque as positions so that values cannot be correlated/guessed. Maybe including something else in the hash might work (position, form load time?).

**Comment by avernet** [comment]
Yes! I meant to say HMAC, not just plain hash (e.g. SHA-1). And this again, if the cost of computing the HMAC for all the necessary item values isn't having too much of a negative impact on performance.

We can also think of other techniques, like a name pool (where we store the values), attached to the document, where we use the index in the name pool when talking with the client.

**Comment by ebruchez** [comment]
Maybe we maybe still use the position, but somehow combine with the Ajax sequence number to differentiate. Not sure if that's feasible.

**Comment by avernet** [comment]
Possible… the server would need to at least keep one past state of the itemset (in addition to the current one). I can't see why would we would need states older than the current or previous, and if that is the case, then this might be simple enough.

**Comment by ebruchez** [comment]
Consider for next release as we are doing itemset work for #4438.

**Comment by ebruchez** [comment]
Discussed today with @avernet.

**Comment by avernet** [comment]
- [ ] Check if fixed by #5986.
